packages/phpstan-rules/src/Rules/RequireStringArgumentInConstructorRule.php

```diff
@@ -0,0 +1,157 @@
+     */
+    private $stringArgPositionsByType = [];
+
+    /**
+     * @var SimpleNameResolver
+     */
+    private $simpleNameResolver;
+
+    /**
+     * @param array<string, array<int>> $stringArgPositionsByType
+     */
+    public function __construct(SimpleNameResolver $simpleNameResolver, array $stringArgPositionsByType = [])
+    {
+        $this->stringArgPositionsByType = $stringArgPositionsByType;
+        $this->simpleNameResolver = $simpleNameResolver;
+    }
+
+    /**
+     * @return string[]
+     */
+    public function getNodeTypes(): array
+    {
+        return [New_::class];
+    }
+
+    /**
+     * @param New_ $node
+     * @return string[]
+     */
+    public function process(Node $node, Scope $scope): array
+    {
+        $errorMessages = [];
+
+        foreach ($this->stringArgPositionsByType as $type => $positions) {
+            if (! $this->isNodeVarType($node, $scope, $type)) {
+                continue;
+            }
+
+            foreach ($node->args as $key => $arg) {
+                if ($this->shouldSkipArg($key, $positions, $arg)) {
+                    continue;
+                }
+
+                $errorMessages[] = sprintf(self::ERROR_MESSAGE, $type, $key);
+            }
+        }
+
+        return $errorMessages;
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(self::ERROR_MESSAGE, [
+            new ConfiguredCodeSample(
+                <<<'CODE_SAMPLE'
+class AnotherClass
+{
+    public function run()
+    {
+        new SomeClass(YetAnotherClass:class);
+    }
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+class AnotherClass
+{
+    public function run()
+    {
+        new SomeClass('YetAnotherClass');
+    }
+}
+CODE_SAMPLE
+                ,
+                [
+                    'stringArgPositionsByType' => [
+                        'SomeClass' => [0],
+                    ],
+                ]
+            ),
+        ]);
+    }
+
+    /**
+     * @param int[] $positions
+     */
+    private function shouldSkipArg(int $key, array $positions, Arg $arg): bool
+    {
+        if (! in_array($key, $positions, true)) {
+            return true;
+        }
+
+        if ($arg->value instanceof String_) {
+            return true;
+        }
+
+        $classConstFetch = $arg->value;
+        if (! $classConstFetch instanceof ClassConstFetch) {
+            return true;
+        }
+
+        return ! $this->simpleNameResolver->isName($classConstFetch->name, 'class');
+    }
+
+    private function isNodeVarType(New_ $constructCall, Scope $scope, string $desiredType): bool
+    {
+        if (trait_exists($desiredType)) {
+            $message = sprintf(
+                'Do not use trait "%s" as type to match, it breaks the matching. Use specific class that is in this trait',
+                $desiredType
+            );
+
+            throw new ShouldNotHappenException($message);
+        }
+
```

Comment:
This should happen on configuration in constuctor. Here is too late, as the validation depends on running the rule with invalid configuration.

In my experience people don't use traits to match types, so personally I would drop it. But feel free to move it to ctor

Reply:
Removed